Asynchronous crash pointer cleanup

burqen opened the pull request:
### WHAT?
Do asynchronous cleanup of crash pointers in `GBPTree` if state of tree is dirty on startup.
Before this cleanup was blocking and done as part of `GBPTree` construction.

### WHY?
When `GBPTree` was only used for `NativeLabelScanStore` having a blocking cleanup procedure was fine because it was only performed on a single tree instance.
Moving forward the `GBPTree` will be used to implement native schema number index and thus there can be several tree instances in store. Up to multiple thousands has been seen in the wild. Having a blocking cleanup procedure would not scale.

### HOW?
- Register cleanup jobs to a `RecoveryCleanupWorkCollector` that will be triggered on `NeoStoreDataSource`s `life.start()`.
- Block checkpointing until cleanup is finished. Checkpoint would make it impossible to later recognise crash pointers that has not yet been cleaned up.
- Block close until cleanup is finished. Because closing the `GBPTree` would also mean closing the underlying `PagedFile` which is pinned by crash pointer cleaner.

Builds on #9231 
First new commit is `Shuffle recovery test load order ...`

Commit:
Shuffle recovery test load order

... to better model the fact that order between transactions
can be different during recovery compared to when they
where first applied, as long as there are no dependencies
(locks) between them.

Commit:
Use readWriteLock in GBPTree

To allow concurrent clean and write.

Commit:
Asynchronous crash pointer cleanup

Before, cleanup of crashed pointers was blocking
and was done as part of GBPTree constructor.

Now, GBPTree just register a cleanup job to
provided RecoveryCleanupWorkCollector that will
be triggered before next checkpoint and cleanup
will be performed asynchronously by some background
thread.

It is safe to have crash pointer cleanup work
concurrently with GBPTree writer because:
- Writer will treat all crashed pointers exactly
  like empty pointers. Ignore when reading and
  overwrite when writing.
- Crash pointer cleaner will overwrite all
  crashed pointers with zeroes, effectively
  turning them into empty pointes.
- Crash pointer execution is not at all affected
  by content of the tree. Just the crashed pointers
  so concurrent structure modifications does not
  matter.

Commit:
Don't force flush as part of recovery

Flushing will implicitly perform a checkpoint on GBPTree.
This will close the generation gap that is necessary to
recognize CRASH pointers in the tree and thus asynchronous
crash gen clean job will not be able to do it's job.

Flushing was the way checkpoint was done before explicit
checkpoint became a thing. And flushing in recovery is an
artifact from that time, it does not do anything useful
other than making recovery take longer time.

Commit:
RecoveryCleanupWorkCollector lifecycle

- Trigger of RecoveryCleanupWorkCollector follow life of
  NeoStoreDataSource.
  Every time GBPTree has registered cleanup jobs,
  RecoveryCleanupWorkCollector needs to be triggered.
  Registering clean jobs (can only) happen in 'init'
  of NeoStoreDataSources life. Therefore we need to
  trigger RecoveryCleanupWorkCollector in 'start' of that
  life.
- Propagate instance to GBPTree
- Don't complete recovery as part of flush
  Was needed previously when cleaning was blocking
  and Recovery did flush everything.

Commit:
TODO remove (Small performance test for CrashGenCleaner)

Commit:
Don't overwrite dirty state and writer use correct lock

- GBPTree leave state 'dirty' on close if it was not cleaned.
- GBPTree.SingleWriter now keeps track of which lock it uses
  and unlock correct one.
- Testing lock for recovery cleanup.

Commit:
Integration test, recovery cleanup should block checkpoint

Commit:
Don't encourage users to delete store files

Commit:
Fix compilation error after rebase

Commit:
Remove temporary benchmark test

Commit:
checkstyle

MishaDemianenko commented on community/neo4j/src/test/java/recovery/RecoveryCleanupIT.java:
```diff
@@ -0,0 +1,194 @@
+    private Barrier.Control recoveryCompleteBarrier;
+    private GraphDatabaseService db;
+    private File storeDir;
+    private ExecutorService executor = Executors.newFixedThreadPool( 2 );
```
executors are not shutdown after the test

MishaDemianenko commented on community/neo4j/src/test/java/recovery/RecoveryCleanupIT.java:
```diff
@@ -0,0 +1,194 @@
+    public void recoveryCleanupShouldBlockCheckpoint() throws Throwable
+    {
+        // GIVEN
+        AtomicReference<Throwable> error = new AtomicReference<>();
```
not quite sure why do we need AtomicReference here to rethrow exception since future.get will do that

burqen replied:
It's because we can not have a throwing function inside the lambda.

MishaDemianenko replied:
```
Future<?> checkpointFuture = executor.submit( (Callable<Void>) () -> {
                checkpoint( db );
                return null;
            } );
```
for example will allow you to get future with checkpoint inside :) but this is also fine it can be like it is now

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTree.java:
```diff
@@ -236,13 +238,7 @@ default void startupState( boolean clean )
-     * guard so that only one thread can have it at any given time.
-     */
-    private final AtomicBoolean writerTaken = new AtomicBoolean();
+    private final ReadWriteLock writerCheckpointMutex = new StampedLock().asReadWriteLock();
```
Is there is any reason why you choose to use stamped lock here? (you do not want reentrancy or smth else?)

burqen replied:
Because the thread that take lock in `cleanCrashedPointers` is not the same thread that will run the clean job and release the lock in the end.

MishaDemianenko replied:
now quite sure, I understand what you mean. Let me be more specific: does this lock needs to be StampedLock or it can be `ReentrantReadWriteLock` as well?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTree.java:
```diff
@@ -308,6 +304,21 @@ default void startupState( boolean clean )
+    /**
+     * If clean job fails, this throwable is set to the cause.
+     */
+    private volatile Throwable failedRecoveryCause;
```
I think it should not be part of GBPTree, if we already have an abstraction that collects and execute recovery tasks then it should be responsible for providing its status, there is no reason to inject status variables inside of jobs

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GBPTree.java:
```diff
@@ -972,12 +1001,58 @@ public String toString()
+            {
+                if ( !success )
+                {
+                    closeCursor();
```
why not just call close here?

MishaDemianenko commented on community/index/src/main/java/org/neo4j/index/internal/gbptree/GroupingRecoveryCleanupWorkCollector.java:
```diff
@@ -0,0 +1,52 @@
+ * Collects recovery cleanup work to be performed and {@link #run() runs} them all as one job,
+ * each cleanup job sequentially one after the other.
+ */
+public class GroupingRecoveryCleanupWorkCollector implements RecoveryCleanupWorkCollector
```
Can you please clarify how this should be used in relation to jobs that will be added from other threads and in relation to lifecycle restarts?
What is the use of started variable in that context?

MishaDemianenko reviewed (commented):
I have couple of comment-questions mostly, otherwise first review done

Commit:
Close executor

Commit:
Don't share state between GBPTree and CleanupJob

CleanupJob take care of its own state and GBPTree ask it
when needed.

Commit:
Use explicit close if unsuccessful initialize

Commit:
Better javadoc for RecoveryCleanupWorkCollector

Commit:
Fix compilation error after rebase

burqen commented:
@MishaDemianenko Comments addressed 🎉

Commit:
Fix checkstyle

MishaDemianenko commented:
@burqen couple of comments to my previous comments and as we just talk - lifecycle of `GroupingRecoveryCleanupWorkCollector` still looks like a problem

Commit:
RecoveryCleanupWorkCollector lifecycle

RecoveryCleanupWorkCollector now extends Lifecycle.

Important details:
 - Register collector to life before any indexes are registered
   because init will clear eventual state (in case of restart) and
   registered cleanup job would be lost.
 - Responsibility of scheduling cleanup jobs is transferred
    from outside trigger (previously created as part of
    buildRecovery in NeoStoreDataSource) to collector itself.

burqen commented:
@MishaDemianenko How do you like this new lifecycle?

Commit:
Checkstyle

Commit:
Move JobScheduler to separate package in common

MishaDemianenko commented:
@burqen will merge on green

Commit:
Fix javadoc

Commit:
JobSchedulerAdapter now live in test

Commit:
Run cleanup jobs sequentially instead of in parallel

Each cleanup job try to use all available processors.
Running them in parallel will only increase the amount
of context switching without utilizing more processors.

burqen commented:
@MishaDemianenko Updated, please note that there are a few new commits addressing comments from @tinwelint (that he left in PR #9362 )

Commit:
Adjust test to new behaviour

Jobs are not longer scheduled one by one
but rather as a single job running all
jobs sequentially. Test now reflect this.
